src/loader/lib/system.js

```diff
@@ -0,0 +1,297 @@
+      (href.protocol || href.authority || href.pathname ? href.search : (href.search || base.search)) +
+      href.hash;
+  }
+
+  var fetchTextFromURL;
+  if (typeof XMLHttpRequest != 'undefined') {
+    fetchTextFromURL = function(url, fulfill, reject) {
+      var xhr = new XMLHttpRequest();
+      var sameDomain = true;
+      var doTimeout = false;
+      if (!('withCredentials' in xhr)) {
+        // check if same domain
+        var domainCheck = /^(\w+:)?\/\/([^\/]+)/.exec(url);
+        if (domainCheck) {
+          sameDomain = domainCheck[2] === window.location.host;
+          if (domainCheck[1])
+            sameDomain &= domainCheck[1] === window.location.protocol;
+        }
+      }
+      if (!sameDomain && typeof XDomainRequest != 'undefined') {
+        xhr = new XDomainRequest();
+        xhr.onload = load;
+        xhr.onerror = error;
+        xhr.ontimeout = error;
+        xhr.onprogress = function() {};
+        xhr.timeout = 0;
+        doTimeout = true;
+      }
+      function load() {
+        fulfill(xhr.responseText);
+      }
+      function error() {
+        var msg = xhr.statusText + ': ' + url || 'XHR error';
+        var err = new Error(msg);
+        err.statusCode = xhr.status;
+        reject(err);
+      }
+
+      xhr.onreadystatechange = function () {
+        if (xhr.readyState === 4) {
+          if (xhr.status === 200 || (xhr.status == 0 && xhr.responseText)) {
+            load();
+          } else {
+            error();
+          }
+        }
+      };
+      xhr.open("GET", url, true);
+
+      if (doTimeout)
+        setTimeout(function() {
+          xhr.send();
+        }, 0);
+
+      xhr.send(null);
+    }
+  }
+  else if (typeof require != 'undefined') {
+    var fs, fourOhFourFS = /ENOENT/;
+    fetchTextFromURL = function(url, fulfill, reject) {
+      if (url.substr(0, 5) != 'file:')
+        throw 'Only file URLs of the form file: allowed running in Node.';
+      fs = fs || require('fs');
+      url = url.substr(5);
+      if (isWindows)
+        url = url.replace(/\//g, '\\');
+      return fs.readFile(url, function(err, data) {
+        if (err) {
+          // Mark this error as a 404, so that the npm extension
+          // will know to retry.
+          if(fourOhFourFS.test(err.message)) {
+            err.statusCode = 404;
+          }
+
+          return reject(err);
+        } else {
+          fulfill(data + '');
+        }
+      });
+    }
+  }
+  else if(typeof fetch === 'function') {
+    fetchTextFromURL = function(url, fulfill, reject) {
+      fetch(url).then(function(resp){
+        return resp.text();
+      }).then(function(text){
+        fulfill(text);
+      }).then(null, function(err){
+        reject(err);
+      });
+    }
+  }
+  else {
+    throw new TypeError('No environment fetch API available.');
+  }
+
+  class SystemLoader extends __global.LoaderPolyfill {
```

Comment:
@matthewp I wonder if we should check in a transpiled version of this file, to get rid of the grunt task that uses `grunt-esnext` it seems to be not deprecated and it made me use an older version of grunt....

or we could maybe merge this thing and open a new issue. Tests are looking good.